core/runtime/src/main/java/io/quarkus/runtime/configuration/Substitutions.java

```diff
@@ -97,4 +115,334 @@ public byte[] getClassBytes() {
+                public Enumeration<?> propertyNames() {
+                    return properties.propertyNames();
+                }
+
+                @Override
+                public Set<String> stringPropertyNames() {
+                    return properties.stringPropertyNames();
+                }
+
+                @Override
+                public void list(final PrintStream out) {
+                    ensureFullyInitialized();
+                    properties.list(out);
+                }
+
+                @Override
+                public void list(final PrintWriter out) {
+                    ensureFullyInitialized();
+                    properties.list(out);
+                }
+
+                @Override
+                public int size() {
+                    return properties.size();
+                }
+
+                @Override
+                public boolean isEmpty() {
+                    return properties.isEmpty();
+                }
+
+                @Override
+                public Enumeration<Object> keys() {
+                    return properties.keys();
+                }
+
+                @Override
+                public Enumeration<Object> elements() {
+                    ensureFullyInitialized();
+                    return properties.elements();
+                }
+
+                @Override
+                public boolean contains(final Object value) {
+                    ensureFullyInitialized();
+                    return properties.contains(value);
+                }
+
+                @Override
+                public boolean containsValue(final Object value) {
+                    ensureFullyInitialized();
+                    return properties.containsValue(value);
+                }
+
+                @Override
+                public boolean containsKey(final Object key) {
+                    return properties.containsKey(key);
+                }
+
+                @Override
+                public Object get(final Object key) {
+                    if (key instanceof String) {
+                        initializeLazyValue((String) key);
+                    }
+                    return properties.get(key);
+                }
+
+                @Override
+                public synchronized Object put(final Object key, final Object value) {
+                    if (key instanceof String) {
+                        initializeLazyValue((String) key);
+                    }
+                    return properties.put(key, value);
+                }
+
+                @Override
+                public synchronized Object remove(final Object key) {
+                    if (key instanceof String) {
+                        initializeLazyValue((String) key);
+                    }
+                    return properties.remove(key);
+                }
+
+                @Override
+                public synchronized void putAll(final Map<?, ?> t) {
+                    properties.putAll(t);
```

Comment:
I'm surprised you handle the initialization in `put` and not here. What's the difference.

Reply:
`put` returns you the old element, so we need to have it initialized if it is lazy.

Reply:
Ah yes, makes sensé, thanks.